### PR TITLE
Signup: AB test Site Title step

### PR DESCRIPTION
**Pull request by bisko**
This PR allows for AB test of a new Signup step, in this case the `Site Title` step.

It adds the ability to inject a step during Signup and not have the steps predefined in the beginning, before the flow has started.

How to test:

1. Checkout branch or use the Calypso.live link below.
2. Start Signup in the `main` flow: `/start/`
3. Go through Signup and see if the Site Title step appears
4. Finish Signup and see if the step is working properly, i.e. the newly created site gets the Site Title you set.
5. Reset and start from 2 :) 
6. Test a variation of the AB test without the Site Title step. This would be `hideSiteTitleStep`. 
7. In the case where the Site Title step is missing or you leave it blank/skip it, the site's tile should be set to a localized version of `Site Title`
8. Also try resuming signup at any point during signup to see how it's behaving. There shouldn't be any change to the behavior from what you were seeing.

Several things to note here:

* It would probably break if you start in the `showSiteTitleStep` variation and then later change it to `hideSiteTitleStep` variation, after you have finished the step. If you ever happen to break signup, just clear out `localStorage` and start fresh.
* The variation is set to 95% to `showSiteTitleStep`. This is just for the test and will be updated before release. Percentages TBD.

TODO:

- [x] Unit tests
- [x] Disable AB variation assignment of new steps to logged in users

cc @michaeldcain @meremagee

### CODE REVIEW

File: client/signup/config/flows.js [hunk 1]
```diff
@@ -238,11 +239,72 @@ const Flows = {
...
+	 *
+	 * This gives the option to set the AB variation as late as possible in the
+	 * signup flow.
+	 *
+	 * Currently only the `main` flow is whitelisted.
+	 *
+	 * @param {String} flowName The current flow
+	 * @param {String} stepName The step that is being completed right now
+	 */
+	preloadABTestVariationsForStep( flowName, stepName ) {
```
Comment (rralian) [inline client/signup/config/flows.js]: I know this is still in-progress and I'm looking early. But it seems like we can simplify this a bit. Rather than injecting the new step in this `preloadABTestVariationsForStep()` method, we could just use `abtest()` to set and return the necessary variation in `getABTestFilteredFlow()`. That way you don't have to mirror the logic in two places. You just need to add a parameter to `getFlow()` to pass in the current stepname so it's available to pass into `Flows.getABTestFilteredFlow()`.

Comment (bisko) [inline client/signup/config/flows.js]: The usage of `getABTestVariation` in `getABTestFilteredFlow` is intentional.

The flow should not be modified until the AB test variation is assigned to the user. The assignment should happen at the latest possible moment, which means after the previous step is completed and before the next one starts. 

`getFlow` and subsequently `getABTestFilteredFlow` is being called since the very early moments of signup initialization and this will force the AB test variants to be assigned even before the first step. If users are assigned this early, the test data may be diluted by the users who don't even reach the step we're testing. 

On the other hand `preloadABTestVariationsForStep` is called after every step (or rather before loading a step). This way we can set the AB variation in this exact moment and not earlier.

With this way of setting the logic, `getFlow` is strictly only giving out the current flow that should be and it will get modified by explicit calls to update the AB variations in `preloadABTestVariationsForStep`. I look at this more like keeping the logic separate, so `getFlow` doesn't get too complicated and black-magic-y.

Of course the whole thing may not make any sense :) I would love to hear thoughts on the above.

Comment (rralian) [inline client/signup/config/flows.js]: > The flow should not be modified until the AB test variation is assigned to the user.

Totally agreed. But you're calling both `flows.preloadABTestVariationsForStep()` and `flows.getFlow()` each time you load the next step. You could just as easily inject the step inside `flow.getFlow()` so you only need to worry about it in one spot... you just need to check the flowname and the stepname before assigning a variant and injecting the dynamic step, just like you're doing in `preloadABTestVariationsForStep()`. It requires passing through the stepname into `getFlow()` to make it available in `getABTestFilteredFlow()` but I think that's preferable to creating a separate method just for injecting the step, and having to maintain branching logic in two places.

One other thing, in case a user navigates back and forth through the steps several times you should check if the new step already exists before injecting it. I think that would be simpler and a little more clear than what you're doing with `.resumingFlow`.

Comment (bisko) [inline client/signup/config/flows.js]: It does make sense to do it in one place :) I updated the code to do that, now the `preloadABTestVariationsForStep` is being called from within `getFlow`, so everything is in one place. 

> One other thing, in case a user navigates back and forth through the steps several times you should check if the new step already exists before injecting it. 

Thanks for pointing that out. I added a check to make sure the step is not already added before adding it in the flow. The `Flows` "class" will undergo a bit of refactoring when I have the time to get to it, to make adding/removing/replacing steps more streamlined than it is now. 

> I think that would be simpler and a little more clear than what you're doing with .resumingFlow.

`resumingFlow` is not related to going back/forth between the steps themselves, but when you start signup, finish some steps, leave and then come back later (e.g. browser crash, page reload, etc.). We don't want to modify an already partly completed flow, or push the test for people that have started before the test went live. Having the check prevents that from happening.

File: client/signup/config/flows.js [hunk 2]
```diff
@@ -238,11 +252,82 @@ const Flows = {
...
+	 *
+	 * Useful when testing new steps in the signup flows.
+	 *
+	 * Example usage: Inject or remove a step in the flow if a user is part of an ABTest.
+	 *
+	 * @param {String} flowName The current flow name
+	 * @param {Object} flow The flow object
+	 */
+	getABTestFilteredFlow( flowName, flow ){
+		const updatedFlow = Object.assign( {}, flow );
```
Comment (mcsf) [inline client/signup/config/flows.js]: Two notes:

- Minor: this assignment is performed even when `flowName !== 'main'`, due to how the function is structured.
- Major: `assign` performs shallowly, meaning that `flow.steps === updatedFlow.steps` even though `flow !== updatedFlow`. 😱

Comment (bisko) [inline client/signup/config/flows.js]: > Major: assign performs shallowly, meaning that flow.steps === updatedFlow.steps even though flow !== updatedFlow. 😱

😥 (disappointed but relieved face - such a proper description). Thanks for catching that, today I learned! :) Updated it with `lodash`'s `cloneDeep` to make sure it's properly cloned.

> Minor: this assignment is performed even when flowName !== 'main', due to how the function is structured.

It definitely makes more of a performance sense to put the expensive clone inside the whitelist check, to make sure there aren't any extra calls to it.

Comment (mcsf) [inline client/signup/config/flows.js]: > Updated it with `lodash`'s `cloneDeep` to make sure it's properly cloned.

That'll do it. :) Also, it's worth knowing that `cloneDeep` is a bit of a blunt instrument which, depending on the situation, may add some cost. Not saying that's the case here, but for the sake of example #6409 shows a situation where a manual copy was preferred — greater control, explicitness, and perf at the cost of a longer and more tedious construction.

Comment (mcsf) [inline client/signup/config/flows.js]: Oh, and as stated in that PR's comments, `react-addons-update` is a viable (and IMO quite nice) alternative.

File: client/signup/config/flows.js (hunk 1 above)
Comment (rralian) [inline client/signup/config/flows.js]: > It does make sense to do it in one place :) I updated the code to do that, now the preloadABTestVariationsForStep is being called from within getFlow, so everything is in one place.

My point was that you don't need both of those methods. You can remove `Flows.preloadABTestVariationsForStep()` and just put your filtering logic (am I in "main" flow and "survey" step?) in `Flows.getABTestFilteredFlow()`. You don't need to assign the variant separately from checking it. `abtest()` returns the variant, so you shouldn't typically need to use `getABTestVariation()`.

File: client/signup/config/flows.js [hunk 3]
```diff
@@ -238,11 +250,84 @@ const Flows = {
...
+		 * E.g. A user is resuming signup from before the test was added. There is no need
+		 * to add a step somewhere back in the line.
+		 */
+		if ( Flows.resumingFlow ) {
+			return;
+		}
+
+		if ( 'main' === flowName ) {
+			if ( 'survey' === stepName ) {
+				abtest( 'siteTitleStep' );
```
Comment (coreh) [inline client/signup/config/flows.js]: Right now I think it's fine to have this hardcoded here, but eventually we should move this to the `flows` object defined earlier in the file.

The entries could then end up like this:

```js
const flows = {

	// ...

	main: {
 		steps: [ 'survey', 'design-type', 'themes', 'domains', 'plans', 'survey-user' ],
 		abtests: { 'survey': [ 'siteTitleStep' ] },

 		// ...

 	},

	// ...

};
```

File: client/signup/config/flows.js [hunk 4]
```diff
@@ -238,11 +250,84 @@ const Flows = {
...
+	 * @param {Object} flow The flow object
+	 */
+	getABTestFilteredFlow( flowName, flow ){
+		/**
+		 * Filter according to running ABTests
+		 */
+
+		// Only do this on the main flow
+		if ( 'main' === flowName ) {
+			const updatedFlow = cloneDeep( flow );
```
Comment (coreh) [inline client/signup/config/flows.js]: Nitpick, but I'd rather move the cloning into `insertStepIntoFlow()`. That way all the parameters are immutable and it's pure functional.

File: client/signup/main.jsx [hunk 5]
```diff
@@ -240,6 +240,9 @@ const Signup = React.createClass( {
 	},
 
 	resumeProgress() {
+		// Update the Flows object to know that the signup flow is being resumed.
+		flows.resumingFlow = true;
```
Comment (coreh) [inline client/signup/main.jsx]: This is good to have, might be useful for other things too.

File: client/signup/config/flows.js [hunk 6]
```diff
@@ -238,11 +250,84 @@ const Flows = {
...
+			return updatedFlow;
+		}
+
+		return flow;
+	},
+
+	insertStepIntoFlow( afterStep, stepName, flow ) {
+		if ( -1 === flow.steps.indexOf( stepName ) ) {
+			const afterStepIndex = flow.steps.indexOf( afterStep );
+			flow.steps.splice( afterStepIndex + 1, 0, stepName );
```
Comment (coreh) [inline client/signup/config/flows.js]: Instead of using the A/B test variants to determine whether to add an extra step, could we use it to determine whether to remove an existing step? That way:

1. We don't need the `afterStep` parameter.
2. The order of the steps goes in the `flows` object defined earlier (declarative), 
3. We make it possible to add optional A/B test steps to the last position, if we ever need that in the future.

We could name the method `flowOmittingStep()` or something similar. 😆

File: client/signup/config/flows.js [hunk 7]
```diff
@@ -226,8 +224,22 @@ const Flows = {
...
+	 *
+	 * `currentStepName` is the current step in the signup flow. It is used
+	 * to determine if any AB variations should be assigned after it is completed.
+	 * Example use case: To determine if a new signup step should be part of the flow or not.
+	 *
+	 * @param {String} flowName The name of the flow to return
+	 * @param {String} currentStepName The current step. See description above
+	 * @returns {Object} A flow object
+	 */
+	getFlow( flowName, currentStepName = '' ) {
```
Comment (coreh) [inline client/signup/config/flows.js]: Should we rename this method to make it more explicit that the returned flow might change as the user progresses through the steps? Like `getModifiedFlow` or something

File: client/signup/test/utils.js [hunk 8]
```diff
@@ -34,7 +34,13 @@ describe( 'utils', function() {
 
 	before( () => {
 		flows = require( 'signup/config/flows' );
+
 		sinon.stub( flows, 'getFlows' ).returns( mockedFlows );
+		sinon.stub( flows, 'preloadABTestVariationsForStep', ()=>{} );
```
Comment (coreh) [inline client/signup/test/utils.js]: Was unsure for a moment on whether `() => {}` would return `{}` or `undefined` 😅

File: client/lib/abtest/active-tests.js [hunk 9]
```diff
@@ -113,5 +113,13 @@ module.exports = {
 			hideVerticalScreenshots: 50,
 		},
 		defaultVariation: 'hideVerticalScreenshots',
+	},
+	siteTitleStep: {
+		datestamp: '20160928',
+		variations: {
+			showSiteTitleStep: 95,
+			hideSiteTitleStep: 5,
```
Comment (bisko) [inline client/lib/abtest/active-tests.js]: These are reversed only until the code gets ready to merge, so we can test the variation and adding of step.

I will revert that before merging.

File: client/signup/test/utils.js (hunk 8 above)
Comment (mcsf) [inline client/signup/test/utils.js]: `undefined` :), though when in doubt use `noop` from `lodash`

File: client/signup/config/flows.js (hunk 4 above)
Comment (mcsf) [inline client/signup/config/flows.js]: Agreed, or even:

```diff
    getABTestFilteredFlow( flowName, flow ){
        /**
         * Filter according to running ABTests
         */

        // Only do this on the main flow
        if ( 'main' === flowName ) {
-           const updatedFlow = cloneDeep( flow );
-
            if ( getABTestVariation( 'siteTitleStep' ) === 'showSiteTitleStep' ) {
-               Flows.insertStepIntoFlow( 'survey', 'site-title', updatedFlow );
+               return Flows.insertStepIntoFlow( 'survey', 'site-title', flow );
            }

-           return updatedFlow;
        }

        return flow;
    },

    insertStepIntoFlow( afterStep, stepName, flow ) {
        if ( -1 === flow.steps.indexOf( stepName ) ) {
+           const steps = flow.steps.slice(); // shallow copy
            const afterStepIndex = flow.steps.indexOf( afterStep );
            steps.splice( afterStepIndex + 1, 0, stepName );
+           return {
+               ...flow,
+               steps,
+           };
        }
+       return flow;
    }
```

File: client/signup/config/flows.js (hunk 3 above)
Comment (bisko) [inline client/signup/config/flows.js]: Yes, I  had a similar idea when I started this, but I think this should be a separate PR.

File: client/signup/config/flows.js (hunk 4 above)
Comment (bisko) [inline client/signup/config/flows.js]: I followed @mcsf's suggestion. It makes more sense to not use the big guns and copy only the `steps` property.

File: client/signup/config/flows.js (hunk 6 above)
Comment (bisko) [inline client/signup/config/flows.js]: I updated the method signature and switched the params a bit. Also added logic to make sure that `afterStep` is really found if it is provided.

If we stick to adding a step, we can add this step dynamically to other flows too, whenever we decide. For example we can start with the `main` flow and then enable the test for all other flows, without having to touch their configuration.

Comment (bisko) [inline client/signup/config/flows.js]: Forgot to mention that we can do something similar with adding or replacing a step according to an AB test as you suggested above.

File: client/signup/config/flows.js (hunk 1 above)
Comment (bisko) [inline client/signup/config/flows.js]: I deliberately used `getABTestVariation` like it is here, because `abtest` sets the variation when it is called.

`getFlow` is the single point of truth for Signup, that gives the current flow configuration. It is called before every step and at the start of Signup.

For testing new steps, we don't want to set the variation until the last possible moment, which is right before the step (in this case - when the previous one ends). If we set the variation in the beginning it may skew the test with all the people that will drop off Signup even before reaching the step. This will require us to run the test longer, so we can gather  more accurate data. If we set the variation right before the step, the only users that will get it will be users that reach the step and it will more accurately measure the step success and attrition rates. 

That's why when `getFlow` gets called, it modifies the flow to what has been already set (e.g. a user resuming signup after page reload or if they navigated away from signup for some reason), so their progress will be properly loaded. Imagine their confusion if they open up signup and some step they already filled is missing and doesn't get sent to the API.

With this, the `preloadABTestVariationsForStep` method updates the flow exactly when it is needed, i.e. when the step is reached. 

cc @meremagee

File: client/signup/config/flows.js (hunk 3 above)
Comment (coreh) [inline client/signup/config/flows.js]: 👍

File: client/signup/config/flows.js [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /**
  * External dependencies
  */
-import { assign, includes, reject, cloneDeep } from 'lodash';
+import { assign, includes, reject } from 'lodash';
```
Comment (coreh) [inline client/signup/config/flows.js]: Removed `cloneDeep`, as it was no longer being used

File: client/signup/config/flows.js [hunk 11]
```diff
@@ -238,11 +250,106 @@ const Flows = {
 			flow = filterDesignTypeInFlow( flow );
 		}
 
-		return flow;
+		if ( ! user.get() ) {
+			// Preload the AB tests before going forward in the flow.
+			// Only valid for logged out users.
+			Flows.preloadABTestVariationsForStep( flowName, currentStepName );
```
Comment (coreh) [inline client/signup/config/flows.js]: So existing users won't be getting site title if they create new sites?

File: client/signup/config/flows.js [hunk 12]
```diff
@@ -238,11 +250,106 @@ const Flows = {
...
+			const steps = flow.steps.slice();
+			const afterStepIndex = steps.indexOf( afterStep );
+
+			/**
+			 * Only insert the step if
+			 * `afterStep` is empty ( insert at start )
+			 * or if `afterStep` is found in the flow. ( insert after `afterStep` )
+			 */
+			if (afterStepIndex > -1 || '' === afterStep ) {
+				steps.splice( afterStepIndex + 1, 0, stepName );
```
Comment (coreh) [inline client/signup/config/flows.js]: Much better now 👍
